### PR TITLE
fix(export): export less & more correct

### PR DESCRIPTION
**Pull request by emjburns**
Fixes https://github.com/spinnaker/keel/issues/762

Fixing some problems reported with exporting. 

- simplifying some logic for converting objects into spec objects (use object mapper now)
- choose the server group with the largest size to be the "base" (because that's probably the main one)
- allow you to override just capacity(min/max) if you have scaling policies also
- pull the export tests out to their own files for titus and ec2 clusters
- update the tests a bit to add new cases and condense some of the `strikt` logic

### CODE REVIEW

File: keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/resource/ClusterHandler.kt [hunk 1]
```diff
@@ -888,20 +896,20 @@ class ClusterHandler(
...
-      health = Health().exportSpec(),
+      health = Health().toSpecWithoutDefaults(),
       scaling = Scaling(),
       tags = emptyMap()
     )
 
-    val thisSpec = ClusterSpec.ServerGroupSpec(
-      launchConfiguration = launchConfiguration.exportSpec(account, application),
+    val thisSpec = ServerGroupSpec(
+      launchConfiguration = launchConfiguration.exportSpec(account, location.region, application, launchConfiguration.instanceType),
```
Comment (emjburns) [inline keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/resource/ClusterHandler.kt]: we need to pass in the instance type here, otherwise we incorrectly think it is overridden.

File: keel-ec2-plugin/src/test/kotlin/com/netflix/spinnaker/keel/ec2/resource/TestUtils.kt [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+import com.netflix.spinnaker.keel.clouddriver.model.SecurityGroupSummary
+import com.netflix.spinnaker.keel.clouddriver.model.Subnet
+import com.netflix.spinnaker.keel.clouddriver.model.SuspendedProcess
+import com.netflix.spinnaker.keel.clouddriver.model.Tag
+import com.netflix.spinnaker.keel.clouddriver.model.TargetTrackingConfiguration
+import com.netflix.spinnaker.keel.ec2.CLOUD_PROVIDER
+import com.netflix.spinnaker.keel.model.parseMoniker
+import org.apache.commons.lang3.RandomStringUtils
+
+fun ServerGroup.toCloudDriverResponse(
```
Comment (emjburns) [inline keel-ec2-plugin/src/test/kotlin/com/netflix/spinnaker/keel/ec2/resource/TestUtils.kt]: Pulling this out to be used by both export and cluster handler tests. No change in functionality.

File: keel-titus-plugin/src/test/kotlin/com/netflix/spinnaker/keel/titus/resource/TestUtils.kt [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+import com.netflix.spinnaker.keel.api.titus.cluster.moniker
+import com.netflix.spinnaker.keel.clouddriver.model.Placement
+import com.netflix.spinnaker.keel.clouddriver.model.SecurityGroupSummary
+import com.netflix.spinnaker.keel.clouddriver.model.ServiceJobProcesses
+import com.netflix.spinnaker.keel.clouddriver.model.TitusActiveServerGroup
+import com.netflix.spinnaker.keel.clouddriver.model.TitusActiveServerGroupImage
+import com.netflix.spinnaker.keel.model.parseMoniker
+import org.apache.commons.lang3.RandomStringUtils
+
+fun TitusServerGroup.toClouddriverResponse(
```
Comment (emjburns) [inline keel-titus-plugin/src/test/kotlin/com/netflix/spinnaker/keel/titus/resource/TestUtils.kt]: Pulling this out to be used by both export and cluster handler tests. No change in functionality.

File: keel-titus-plugin/src/test/kotlin/com/netflix/spinnaker/keel/titus/resource/TitusClusterHandlerTests.kt [hunk 4]
```diff
@@ -470,103 +433,6 @@ class TitusClusterHandlerTests : JUnit5Minutests {
             .containsDistinctElements()
         }
       }
-      context("export without overrides") {
```
Comment (emjburns) [inline keel-titus-plugin/src/test/kotlin/com/netflix/spinnaker/keel/titus/resource/TitusClusterHandlerTests.kt]: Pulled to their own file, and update.

File: keel-ec2-plugin/src/test/kotlin/com/netflix/spinnaker/keel/ec2/resource/ClusterHandlerTests.kt [hunk 5]
```diff
@@ -444,72 +349,6 @@ internal class ClusterHandlerTests : JUnit5Minutests {
           verify { publisher.publishEvent(ofType<ArtifactVersionDeployed>()) }
         }
       }
-
```
Comment (emjburns) [inline keel-ec2-plugin/src/test/kotlin/com/netflix/spinnaker/keel/ec2/resource/ClusterHandlerTests.kt]: Pulled to their own file, and updated.

File: keel-plugin/src/main/kotlin/com/netflix/spinnaker/keel/plugin/ResourceSpecExportHelper.kt [hunk 6]
```diff
@@ -60,24 +60,3 @@ inline fun <reified T : Any, reified S : Any> buildSpecFromDiff(
     ctor.callBy(params)
   }
 }
-
```
Comment (emjburns) [inline keel-plugin/src/main/kotlin/com/netflix/spinnaker/keel/plugin/ResourceSpecExportHelper.kt]: Removed in favor of using the object mapper

File: keel-titus-plugin/src/test/kotlin/com/netflix/spinnaker/keel/titus/resource/TitusClusterHandlerTests.kt [hunk 7]
```diff
@@ -174,34 +165,6 @@ class TitusClusterHandlerTests : JUnit5Minutests {
     )
   )
 
-  private fun TitusServerGroup.toClouddriverResponse(
```
Comment (emjburns) [inline keel-titus-plugin/src/test/kotlin/com/netflix/spinnaker/keel/titus/resource/TitusClusterHandlerTests.kt]: Pulled to TestUtils

File: keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/resource/ClusterHandler.kt [hunk 8]
```diff
@@ -337,8 +346,7 @@ class ClusterHandler(
     val spec = ClusterSpec(
       moniker = exportable.moniker,
       imageProvider = if (base.buildInfo?.packageName != null) {
-        ArtifactImageProvider(
-          deliveryArtifact = DebianArtifact(name = base.buildInfo.packageName!!))
+        ReferenceArtifactImageProvider(reference = base.buildInfo.packageName.toString())
```
Comment (luispollo) [inline keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/resource/ClusterHandler.kt]: This seems unrelated?

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/exceptions/ExportError.kt [hunk 9]
```diff
@@ -0,0 +1,3 @@
+package com.netflix.spinnaker.keel.exceptions
+
+class ExportError(message: String) : RuntimeException(message)
```
Comment (robfletcher) [inline keel-core/src/main/kotlin/com/netflix/spinnaker/keel/exceptions/ExportError.kt]: Is this translated into an HTTP error on the export API?

File: keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/resource/ClusterHandler.kt (hunk 8 above)
Comment (emjburns) [reply keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/resource/ClusterHandler.kt]: oh it totally is. But I wanted to do it. I can back it out if you care

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/exceptions/ExportError.kt (hunk 9 above)
Comment (emjburns) [reply keel-core/src/main/kotlin/com/netflix/spinnaker/keel/exceptions/ExportError.kt]: Do you think this should be a `500`? this is basically just used because kotlins null safety didn't catch that I throw an exception when there's an empty list, but then finding the first in the list can still be null. I don't think we will ever throw it, and if we don't I think it's a bug on our side.

Comment (robfletcher) [reply keel-core/src/main/kotlin/com/netflix/spinnaker/keel/exceptions/ExportError.kt]: In that case 500 seems right. I was more just wondering if this exception might start getting used for other things in future.

Comment (emjburns) [reply keel-core/src/main/kotlin/com/netflix/spinnaker/keel/exceptions/ExportError.kt]: I don't think it will, but if it does I'll add the exception handling by the controllers.

**Comment by gal-yardeni** [comment]
LGTM! I think it makes sense to choose the server group with the largest size to be the "base" as opposed to just choosing the first one.

**Comment by emjburns** [comment]
I screwed up this PR. New one here: https://github.com/spinnaker/keel/pull/795
